### PR TITLE
Semantic analysis check for USEION variables in CONSTANT block

### PR DESCRIPTION
**Pull request by pramodk**
* Similar to https://github.com/neuronsimulator/nrn/pull/1955
* Add semantic analysis check to avoid declaring ion variables
  in CONSTANT {} block
* Addresses first point in https://github.com/BlueBrain/nmodl/issues/888
  MOD file is glia__dbbs_mod_collection__cdp5__CAM_GoC.mod.
* Updated test

### CODE REVIEW

File: src/visitors/semantic_analysis_visitor.cpp [hunk 1]
```diff
@@ -22,6 +23,28 @@ bool SemanticAnalysisVisitor::check(const ast::Program& node) {
...
+    const auto& with_prop = NmodlType::read_ion_var | NmodlType::write_ion_var;
+
+    const auto& sym_table = node.get_symbol_table();
+    assert(sym_table != nullptr);
+
+    // get all ion variables
+    const auto& variables = sym_table->get_variables_with_properties(with_prop, false);
+
+    /// make sure ion variables aren't redefined in a `CONSTANT` block.
+    for (const auto& var: variables) {
```
Comment (ohm314) [inline src/visitors/semantic_analysis_visitor.cpp:37]:
```suggestion
    const auto& ion_variables = sym_table->get_variables_with_properties(with_prop, false);

    /// make sure ion variables aren't redefined in a `CONSTANT` block.
    for (const auto& var: ion_variables) {
```

**Review by ohm314 — APPROVED** [review]
really minor nitpicking otherwise looks good

File: test/unit/visitor/semantic_analysis.cpp [hunk 2]
```diff
@@ -105,3 +107,18 @@ SCENARIO("Destructor block", "[visitor][semantic_analysis]") {
...
+                SUFFIX cdp4Nsp
+                USEION ca READ cao, cai, ica WRITE cai
+            }
+            CONSTANT { cao = 2  (mM) }
+        )";
+        THEN("Semantic analysis fails") {
+            REQUIRE(run_semantic_analysis_visitor(nmodl_text));
+        }
+    }
+}
```
Comment (iomaganaris) [inline test/unit/visitor/semantic_analysis.cpp:124]: Nitpick.. New line missing at the end of the file

**Review by iomaganaris — APPROVED** [review]
LGTM
Just a minor nitpick

File: test/unit/visitor/semantic_analysis.cpp (hunk 2 above)
Comment (pramodk) [reply test/unit/visitor/semantic_analysis.cpp]: note that new clang-format (`./cmake/hpc-coding-conventions/bin/format`) removed the newlines automatically.

File: src/visitors/semantic_analysis_visitor.cpp [hunk 3]
```diff
@@ -22,6 +23,28 @@ bool SemanticAnalysisVisitor::check(const ast::Program& node) {
     }
     /// -->
 
+    /// <-- This code is for check 3
```
Comment (alkino) [inline src/visitors/semantic_analysis_visitor.cpp:26]: check 4 I guess?

Comment (pramodk) [reply src/visitors/semantic_analysis_visitor.cpp]: Ah yes! I fixed the merge conflicts but didn't realize the above needs to be changed! thanks!

**Pull request merged**